File: medium/1927-sum-game/solution.py
class Solution:
    def sumGame(self, num: str) -> bool:
        """
        1927. Sum Game
        Time: O(n)
        Space: O(1)
        """
        n = len(num)
        half = n // 2

        sum_left = sum_right = 0
        q_left = q_right = 0
        for i, ch in enumerate(num):
            if i < half:
                if ch == "?":
                    q_left += 1
                else:
                    sum_left += int(ch)
            else:
                if ch == "?":
                    q_right += 1
                else:
                    sum_right += int(ch)

        # Odd number of blanks: Alice moves last and can always break equality.
        if (q_left + q_right) % 2 == 1:
            return True

        # Surplus blanks on one side are worth 9 per pair under optimal play.
        return sum_left - sum_right != 9 * (q_right - q_left) // 2

File: medium/1927-sum-game/solution.java
/**
 * 1927. Sum Game
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public boolean sumGame(String num) {
        int n = num.length();
        int half = n / 2;

        int sumLeft = 0, sumRight = 0;
        int qLeft = 0, qRight = 0;
        for (int i = 0; i < n; i++) {
            char ch = num.charAt(i);
            if (i < half) {
                if (ch == '?') {
                    qLeft++;
                } else {
                    sumLeft += ch - '0';
                }
            } else {
                if (ch == '?') {
                    qRight++;
                } else {
                    sumRight += ch - '0';
                }
            }
        }

        // Odd number of blanks: Alice moves last and can always break equality.
        if ((qLeft + qRight) % 2 == 1) {
            return true;
        }

        // Surplus blanks on one side are worth 9 per pair under optimal play.
        return sumLeft - sumRight != 9 * (qRight - qLeft) / 2;
    }
}
